File: C++/virtual-world-simulator/src/Entity/Animals/Human.cpp
#include "Human.h"

Human::Human(World* world, Point position) :
    Animal{ AssetManager::getAssetManager()->getAsset("human"), world, position, 'H', 5, 4 }, newPosition{ position }, abilityTurnedOn{ false }
{
}

Human::Human(World* world, Point position, int lifeSpan, int strength) :
    Animal{ AssetManager::getAssetManager()->getAsset("human"), world, position, 'H', strength, 4, lifeSpan }, newPosition{ position }, abilityTurnedOn{ false }
{
}

Point Human::getNewPosition() const
{
    return this->newPosition;
}

bool Human::checkIfAbilityTurnedOn() const
{
    return this->abilityTurnedOn;
}

void Human::switchAbility(bool value)
{
    this->abilityTurnedOn = value;
}

bool Human::setNewPosition(int keyCode)
{
    int mapWidth = this->world->getWidth();
    int mapHeight = this->world->getHeight();

    switch (keyCode)
    {
        case KeyCodes::UP:
        {
            if (this->position.x == 0)
                break;

            newPosition.x = (this->position.x - 1);
            return true;
        }
        case KeyCodes::DOWN:
        {
            if (this->position.x == mapHeight - 1)
                break;

            newPosition.x = (this->position.x + 1);
            return true;
        }
        case KeyCodes::LEFT:
        {
            if (this->position.y == 0)
                break;

            newPosition.y = (this->position.y - 1);
            return true;
        }
        case KeyCodes::RIGHT:
        {
            if (this->position.y == mapWidth - 1)
                break;

            newPosition.y = (this->position.y + 1);
            return true;
        }
    }

    return false;
}

void Human::update(std::fstream& logFile)
{
    this->move(logFile, newPosition);

    if (!this->isAlive)
        world->endGame();
}

bool Human::collision(std::fstream& logFile, Entity& entity)
{
    if (abilityTurnedOn)
    {
        Entity*** map = world->getMap();
        int currentX = this->position.x;
        int currentY = this->position.y;

        for (int i = currentX - 1; i <= currentX + 1; i++)
        {
            if (i < 0 || i > world->getHeight() - 1)
                continue;

            for (int j = currentY - 1; j <= currentY + 1; j++)
            {
                if (j < 0 || j > world->getWidth() - 1)
                    continue;

                if (map[i][j] == nullptr)
                {
                    map[i][j] = this;
                    map[currentX][currentY] = nullptr;
                    this->position = { i, j };
                    return true;
                }
            }
        }

        return false;
    }

    bool temp = Animal::collision(logFile, entity);
    if (!this->isAlive)
        world->endGame();

    return temp;
}

File: Python/virtual-world-simulator/src/Entity/Animals/Human.py
import pygame
from src.Entity.Animals.Animal import Animal


class Human(Animal):

    def __init__(self, world, position):
        temp = pygame.image.load('assets/human.png')
        scaled = pygame.transform.scale(temp, (world.scale, world.scale))
        super().__init__(scaled, world, position, 'H', 5, 4)
        self._newPosition = (position[0], position[1])
        self._abilityTurnedOn = False

    @property
    def abilityTurnedOn(self):
        return self._abilityTurnedOn

    @abilityTurnedOn.setter
    def abilityTurnedOn(self, value):
        self._abilityTurnedOn = value

    def setNewPosition(self, keyCode):
        mapWidth = self._world.width
        mapHeight = self._world.height

        if keyCode == pygame.K_UP:
            if self._position[0] != 0:
                self._newPosition = (self._position[0] - 1, self._position[1])
                return True

        elif keyCode == pygame.K_DOWN:
            if self._position[0] != mapHeight - 1:
                self._newPosition = (self._position[0] + 1, self._position[1])
                return True

        elif keyCode == pygame.K_LEFT:
            if self._position[1] != 0:
                self._newPosition = (self._position[0], self._position[1] - 1)
                return True

        elif keyCode == pygame.K_RIGHT:
            if self._position[1] != mapWidth - 1:
                self._newPosition = (self._position[0], self._position[1] + 1)
                return True

        return False

    def update(self):
        self.move(self._newPosition)

        if not self._isAlive:
            self._world.gameOver = True

    def collision(self, entity):

        if self._abilityTurnedOn:
            currentX = self._position[0]
            currentY = self._position[1]

            for i in range(currentX - 1, currentX + 2):
                if i < 0 or i > self._world.height - 1:
                    continue

                for j in range(currentY - 1, currentY + 2):
                    if j < 0 or j > self._world.width - 1:
                        continue

                    if self._world.getMapElement(i, j) is None:
                        self.move((i, j))
                        return True

            return False

        return super().collision(entity)

    def kill(self):
        if not self._abilityTurnedOn:
            super().kill()
            self._world.gameOver = True
